cpp/src/arrow/compute/exec/expression_execution_benchmark.cc

```diff
@@ -0,0 +1,56 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/cast.h"
+#include "arrow/compute/exec/expression.h"
+#include "arrow/compute/exec/test_util.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+
+namespace arrow {
+namespace compute {
+
+auto expression =
+    and_(less(field_ref("x"), literal(20)), greater(field_ref("x"), literal(0)));
+
+static void Execution(benchmark::State& state) {
+  const auto array_batches = static_cast<int32_t>(state.range(0));
+  const auto array_size = 10000000 / array_batches;
+
+  ExecContext ctx;
+  auto dataset_schema = schema({
+      field("x", int64()),
+  });
+  ExecBatch input({Datum(ConstantArrayGenerator::Int64(array_size, 5))},
+                  /*length=*/1);
+
+  ASSIGN_OR_ABORT(auto bound, expression.Bind(*dataset_schema));
+  for (auto _ : state) {
+    for (int it = 0; it < array_batches; ++it)
+      ABORT_NOT_OK(ExecuteScalarExpression(bound, input, &ctx).status());
+  }
+  state.SetBytesProcessed(state.iterations() * 8 * array_size);
```

Comment:
Bytes-per-second is not really a measure of overhead, is it?

Reply:
At least this would be more informative as items-per-second, because the byte size of the data is irrelevant here.

Reply:
This is probably my fault as I asked for bytes-per-second in the JIRA issue.  I think items per second is a good compromise.

Reply:
Also, I think this would need to be `state.iterations() * 8 * array_size * array_batches`

Reply:
unless I'm misunderstanding what the bytes processed is counting? I see a single array being processed, so I think items and bytes are directly related.

Reply:
They are directly related and yes, interchangeable, but I think I agree with @pitrou that items/second is probably more likely to be all we need.  It also makes it clearer to the reader what is being tested.  The code we are testing here (kernel dispatch) never actually operates on the arrays themselves.  It is only responsible for passing the arrays to the appropriate kernels.  So the size (or byte width) of the array shouldn't matter too much.

The kernel benchmarks themselves are a different story.  And you will notice that in `compare_benchmark.cc` for example we test the kernel on a wide variety of data types.

Reply:
Ah, thanks for the clarification; that makes a lot of sense.

Reply:
Made the change, now we are using `items per second`